### PR TITLE
Validation Suite (PC Relate)

### PR DESCRIPTION
**Pull request by ravwojdyla**
Depends on #228 (related: #24)

See: https://github.com/pystatgen/sgkit/pull/228#issuecomment-687607928 for context.

Todo:
 - [x] merge #228 
 - [x] freeze all versions of R packages etc
 - [x] stage public data to a GCS bucket
 - [x] configure github actions to allow to read from the GCS bucket

This validation job currently tests sgkit_plink's `read_plink` + sgkit's `pc_relate` on HapMap data.

This is blocked on pystatgen/sgkit-plink#32.

### CODE REVIEW

**Comment by ravwojdyla** [comment]
Validated that the workflow *should* work via [act](https://github.com/nektos/act):

```
act -sGCP_SA_KEY -sGCP_PROJECT_ID -P ubuntu-latest=nektos/act-environments-ubuntu:18.04 -j validation_suite
```

**Review by jeromekelleher — APPROVED** [review]
LGTM. Pretty solid validation.

**Comment by eric-czech** [comment]
Looks good to me too, that gcloud github action is very cool.

It is nagging me though that the last part isn't a unit test.  I can imagine that if we adopt this approach more broadly, we will have bugs we don't know about for up to a week and that it will be hard to debug them since it would take some work to set this up to be interactive.  Or at least I'm imagining that if we accidentally make a change that alters our pc-relate (far more likely than a meaningful change in GENSIS no doubt), I can see myself copying/pasting the `test_same_as_the_reference_implementation` code locally with a copy of that csv to figure out what the difference is.

Do you think it would make more sense @ravwojdyla / @jeromekelleher if this process simply produced the data we validate in unit tests on a schedule rather than being our only automated way of checking against it too?

If we did the same with say REGENIE, I would definitely want an easy way to fire up the debugging context since the comparable `test_same_as_the_reference_implementation` test for it is the most useful one by far and also much more likely to break than the others.

**Comment by jeromekelleher** [comment]
> Do you think it would make more sense @ravwojdyla / @jeromekelleher if this process simply produced the data we validate in unit tests on a schedule rather than being our only automated way of checking against it too?

Saving the CSVs makes sense, I guess, but we do have to take into account how long things need to run for unit tests. There's a balance here between making the validation realistic and the sub-second run times we need for unit tests. On the whole, I think I'd be in favour of more realistic validations even if that does mean we have a lag between some problems showing up. There's also the size of the repo to consider, which already seems to be getting pretty lardy.

There should be some simple way of running these validations locally though, I agree.

**Comment by ravwojdyla** [comment]
> Or at least I'm imagining that if we accidentally make a change that alters our pc-relate (far more likely than a meaningful change in GENSIS no doubt), I can see myself copying/pasting the test_same_as_the_reference_implementation code locally with a copy of that csv to figure out what the difference is.

@eric-czech @jeromekelleher to clear, to run this manually you don't need to copy anything, just follow this [doc](https://github.com/pystatgen/sgkit/blob/c5bc07ae9309e0c004eeb10fa5c2538ef70f421f/validation/gwas/method/pc_relate/README.md):

```
gsutil -u $BILLING_PROJECT cp gs://sgkit-data/validation/hapmap_JPT_CHB_r23a_filtered.zip /tmp/
./run.sh /tmp/hapmap_JPT_CHB_r23a_filtered.zip
```

I think there is space for unit tests and there is space for more integration/validation style tests (for example right now this test is catching the bug that `sgkit_plink` is broken via pystatgen/sgkit-plink#32). And I would argue that we should have both. Unit tests should catch any bugs introduced in `pc_relate`, and this validation is just an extra safety net (+ documentation, that should not go stale because it's being run regularly). Plus if we want nothing stops us from running this more often than weekly (daily? etc)

> Do you think it would make more sense @ravwojdyla / @jeromekelleher if this process simply produced the data we validate in unit tests on a schedule rather than being our only automated way of checking against it too?

The test data in this case is in 100s of MB, we can't stage that in the git repo, we would need to use LFS (which bunch of people, including me, had bad experience with), so we would probably need to stage that on GCS, then we need to fetch data from GCS in unit test, so we probably need to build caching for it as well, and so on, I'm not sure if that is worth it.

> If we did the same with say REGENIE, I would definitely want an easy way to fire up the debugging context since the comparable test_same_as_the_reference_implementation test for it is the most useful one by far and also much more likely to break than the others.

If you run this locally (as documented above, you can definitely connect debugger there). Also I think in the case of REGENIE, your test data is relatively small, so that is fine, but maybe we could add a bigger validation test for REGENIE  and test against GLOW or the cpp implementation.

What do you think @eric-czech @jeromekelleher ?

**Comment by eric-czech** [comment]
> The test data in this case is in 100s of MB, we can't stage that in the git repo

Ah ok I see.

> Plus if we want nothing stops us from running this more often than weekly (daily? etc)

Probably better than randomly sampling to a relatively tiny subset of the same data for a unit test?  I would definitely have a preference for that over trying to fetch big datasets from GCS in the build.  I suppose we can cross that bridge later if setting up a local process ever becomes time consuming.

**Review by eric-czech — APPROVED** [review]

**Comment by ravwojdyla** [comment]
> Probably better than randomly sampling to a relatively tiny subset of the same data for a unit test?

I believe so, and even with sampling, it's not enough to sample the output data in this case, the input is large as well, and if we sample input, we no longer actually validating against "real world data" anymore. I'm happy to change to schedule to daily, will push an update to this PR.

File: validation/gwas/method/pc_relate/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+HapMap data.
+
+This code is scheduled as part of the Github Actions CI.
+
+To run manually:
+
+```bash
+gsutil -u $BILLING_PROJECT cp gs://sgkit-data/validation/hapmap_JPT_CHB_r23a_filtered.zip /tmp/
+./run.sh /tmp/hapmap_JPT_CHB_r23a_filtered.zip
+```
```
Comment (tomwhite) [inline validation/gwas/method/pc_relate/README.md:12]: Might be worth adding a comment about requester-pays and the data size to the README.

**Review by tomwhite — APPROVED** [review]
Excellent +1

File: validation/gwas/method/pc_relate/README.md (hunk 1 above)
Comment (ravwojdyla) [reply validation/gwas/method/pc_relate/README.md:16]: Fixed

**Comment by ravwojdyla** [comment]
Good to merge?

**Comment by jeromekelleher** [comment]
Three +1s here, so marking for merge.

**Comment by ravwojdyla** [comment]
@jeromekelleher I'm sorry you need to go through this, but his one also changes github workflows -> needs manual merge please.

**Comment by jeromekelleher** [comment]
@mergifyio update

**Comment by jeromekelleher** [comment]
Ok, thanks @ravwojdyla - I'll merge once CI passes again after update

**Pull request merged**